File: app/src/main/java/info/nukoneko/cuc/android/kidspos/viewmodel/StoreListViewModel.java
package info.nukoneko.cuc.android.kidspos.viewmodel;

import android.content.Context;
import android.databinding.ObservableInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import java.util.List;

import info.nukoneko.cuc.android.kidspos.KidsPOSApplication;
import info.nukoneko.cuc.android.kidspos.entity.Store;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;

public final class StoreListViewModel implements ViewModel {

    public ObservableInt progressVisibility;
    public ObservableInt recyclerViewVisibility;
    public ObservableInt errorButtonVisibility;

    @Nullable
    private Context mContext;
    @Nullable
    private DataListener mListener;

    public StoreListViewModel(@NonNull Context context, @NonNull DataListener listener) {
        mContext = context;
        mListener = listener;

        progressVisibility = new ObservableInt(View.GONE);
        recyclerViewVisibility = new ObservableInt(View.GONE);
        errorButtonVisibility = new ObservableInt(View.GONE);
    }

    @Override
    public void destroy() {
        mContext = null;
        mListener = null;
    }

    public void onActivityCreated() {
        loadStores();
    }

    private void loadStores() {
        progressVisibility.set(View.VISIBLE);
        recyclerViewVisibility.set(View.GONE);
        errorButtonVisibility.set(View.GONE);

        if (mContext == null) return;
        KidsPOSApplication app = KidsPOSApplication.get(mContext);
        if (app == null) return;
        app.getStoreRepository().fetchStores()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(app.defaultSubscribeScheduler())
                .subscribe(new Consumer<List<Store>>() {
                    @Override
                    public void accept(List<Store> stores) throws Exception {
                        if (mListener != null) mListener.onStoresChanged(stores);
                        progressVisibility.set(View.GONE);
                        if (!stores.isEmpty()) {
                            recyclerViewVisibility.set(View.VISIBLE);
                        } else {
                            errorButtonVisibility.set(View.VISIBLE);
                        }
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        if (mListener != null) mListener.onError("リストの取得に失敗しました");
                        progressVisibility.set(View.GONE);
                        errorButtonVisibility.set(View.VISIBLE);
                    }
                });
    }

    public interface DataListener {
        void onStoresChanged(List<Store> stores);

        void onError(String message);
    }
}

File: app/src/main/java/info/nukoneko/cuc/android/kidspos/ui/main/storelist/StoreListViewModel.kt
package info.nukoneko.cuc.android.kidspos.ui.main.storelist

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.view.View
import info.nukoneko.cuc.android.kidspos.KidsPOSApplication
import info.nukoneko.cuc.android.kidspos.entity.Store
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class StoreListViewModel(application: Application) : AndroidViewModel(application) {

    private val data = MutableLiveData<List<Store>>()
    fun getData(): LiveData<List<Store>> = data

    private val progressVisibility = MutableLiveData<Int>()
    fun getProgressVisibility(): LiveData<Int> = progressVisibility

    private val recyclerViewVisibility = MutableLiveData<Int>()
    fun getRecyclerViewVisibility(): LiveData<Int> = recyclerViewVisibility

    private val errorButtonVisibility = MutableLiveData<Int>()
    fun getErrorButtonVisibility(): LiveData<Int> = errorButtonVisibility

    var listener: Listener? = null

    private val compositeDisposable = CompositeDisposable()

    init {
        progressVisibility.value = View.GONE
        recyclerViewVisibility.value = View.GONE
        errorButtonVisibility.value = View.GONE
    }

    override fun onCleared() {
        listener = null
        compositeDisposable.clear()
        super.onCleared()
    }

    fun onResume() {
        fetchStores()
    }

    private fun fetchStores() {
        val app = KidsPOSApplication[getApplication()] ?: return

        progressVisibility.postValue(View.VISIBLE)
        recyclerViewVisibility.postValue(View.GONE)
        errorButtonVisibility.postValue(View.GONE)

        val disposable = app.storeManager.fetchStores()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(app.getDefaultSubscribeScheduler())
                .subscribe({ stores ->
                    onFetchStoresSuccess(stores)
                }, {
                    onFetchStoresFailure(it)
                })
        compositeDisposable.add(disposable)
    }

    private fun onFetchStoresSuccess(stores: List<Store>) {
        data.postValue(stores)
        progressVisibility.postValue(View.GONE)
        if (!stores.isEmpty()) {
            recyclerViewVisibility.postValue(View.VISIBLE)
        } else {
            errorButtonVisibility.postValue(View.VISIBLE)
        }
    }

    private fun onFetchStoresFailure(error: Throwable) {
        listener?.onError(error.localizedMessage)
        progressVisibility.postValue(View.GONE)
        errorButtonVisibility.postValue(View.VISIBLE)
    }

    interface Listener {
        fun onError(message: String)
    }
}
